RFC: Add basic support for local and remote CAN

stoupa-cz opened the pull request:
**Description**

This is a draft of basic CAN support. It is in RFC, as there are open questions, missing docs and tests, and moreover, it needs more testing. ~~We plan to use it to remotely access the satellite's telemetry buses during development.~~ It is tested on our local branch to remotely access the satellite's telemetry buses during development.

Local and remote CAN resources are added. Additionally, a simple CAN driver is added. The driver has very limited capabilities. For local use, it can set up the interface and tell you the interface name. For remote use, it establishes a transparent tunnel using cannelloni. With it, you can use the local virtual CAN interface. All the traffic is routed to the real interface on the other end.

Be aware, as cannelloni states in its doc:

> cannelloni is not suited for production deployments. Use it only in environments where packet loss is tolerable. There is no guarantee that CAN frames will reach their destination at all and/or in the right order.

~~I tested in the lab only. We will test it in production during September/October.~~

Draft version (without `labgrid-can-interface`) is being tested in production for ~two months.

~~I see the main blocker: it requires root privileges on both sides (exporter and client; even for local use). What is the preferred way to handle privileged access? May we use `labgrid-raw-interface`?~~ Helper `labgrid-can-interface` has been added.

**Checklist**
- [x] Solve the need for root privileges
- [x] Test initial implementation
- [ ] More robust handling of cannelloni processes
- [ ] CAN configuration via labgrid or e.g., systemd-networkd, [see the discission](https://github.com/labgrid-project/labgrid/pull/1719#issuecomment-3668743001)
- [ ] Fix remaining TODOs in the code
- [ ] Rebase on master
- [ ] Documentation
  - [ ] The arguments and description in doc/configuration.rst have been updated 
  - [ ] Add a section on how to use the feature to doc/usage.rst
  - [ ] Man pages have been regenerated
- [ ] Write tests
- [ ] Test in production

Reference issues: #1287 and #1289 

This is a part of my work at VZLU AEROSPACE.

stoupa-cz committed:
resource/udev: filter CAN devices in the network interfaces

The CAN devices are matched in the network interfaces. The CAN device is
not a typical network interface, so filter it. A new resource, CANPort,
will be introduced to cover the CAN devices.

Signed-off-by: Tomas Novotny <tomas@novotny.cz>

stoupa-cz committed:
resource: add support for CAN connected locally

Add a simple CAN resource. It may be used to determine the interface
name by udev matching.

Signed-off-by: Tomas Novotny <tomas@novotny.cz>

stoupa-cz committed:
add NetworkCANPort resource and export

Add an exporter part for the remotely accessible CAN. As discussed, the
cannelloni is used as the underlying tunnel.

According to the cannelloni documentation, it should be used only in
environments where packet loss is tolerable. There is no guarantee that
CAN frames will reach their destination at all and/or in the right
order.

Signed-off-by: Tomas Novotny <tomas@novotny.cz>

stoupa-cz committed:
driver: add CAN driver

The driver can connect to both local and remote CAN ports. For local
ports, it sets up the interface. The driver is also helpful for
identifying the interface name using udev. For remote ports, it creates
a local vcan interface and establishes a connection to the exported
resource through cannelloni.

According to the cannelloni documentation, the speed of the network
interface is limited by tc to prevent packet losses.

XXX: See the code (a few things to fix).

Signed-off-by: Tomas Novotny <tomas@novotny.cz>

hundeboll commented:
Hi @stoupa-cz 

Thanks for working on this. I've been experimenting with `[socketcand](https://github.com/linux-can/socketcand)`. It has the same limitation as `cannelloni` (root on both ends for `vcan` setup), but the protocol is pretty simple, so the client side can be implemented as python functions in the labgrid driver. This should be enough for tests with limited CAN traffic.

stoupa-cz committed:
helpers/labgrid-can-interface: introduce CAN interface helper

The helper focuses on the CAN interface. All actions (except `add`) may
be run only on interfaces of type CAN.

It is heavily inspired by `labgrid-raw-interface`. Two subcommands are
currently supported. The `ip` subcommand manages the interface, while
the `tc` subcommand throttles the interface according to the Cannelloni
documentation.

TODO: documentation

Signed-off-by: Tomas Novotny <tomas@novotny.cz>

stoupa-cz commented:
> I tried running an exporter with your PR, where a `USBCANPort` is exported as an `NetworkCANPort`. On the client, I implemented the `cannelloni` protocol in the driver ([can.py](https://github.com/user-attachments/files/24166243/can.py)).
> 
> It works quite well for our purpose.
> 
> One thing I miss from `socketcand` is the ability to subscribe to certain CAN IDs, so I added (very) simple filtering in the driver.

It seems we have different needs. For us, a client-side "real" `vcan` interface is necessary. Fortunately, we have a configurable driver, so any client (e.g., your cannelloni protocol reimplementation) can be selected. For the sake of simplicity, I’d prefer to continue using the "native" cannelloni approach.

BTW, what is your use case? Do you want to replay some traffic?
 
> And again: It might make sense to remove the interface configuration, and leave that to e.g. `systemd-networkd`?

On exporter side, the interface is controlled (e.g., down, set-bitrate, up) when the place is acquired. The bitrate configuration comes from the expoter configuration. It is similar on client when the driver is activated, although the bitrate is loaded from the exporter's configuration. Which interface configuration managed by `systemd-networkd` are you referring to?

hundeboll commented:
I agree. Just let the CANPort resource default to no bitrate setting, and only run `ip link ...` commands if a bitrate is configured. Having a default is probably bad anyways, as we cannot guess the bitrate reliably.

I think we should focus on getting the resource merged, as it's easy to write custom drivers in the client context. Then we can discuss the details of the driver later.

stoupa-cz commented:
`vxcan` in netns might be a replacement, see https://github.com/labgrid-project/labgrid/issues/1289#issuecomment-4213516277 for more details.
